src/routes.rs

```diff
@@ -82,3 +84,51 @@ pub async fn create_account(
 
     Ok(warp::reply::json(&UserWithoutSecrets::from(new_user)))
 }
+
+/// Collect money for a user by multiplying their production by the time since last collection and then adding it to their balance
+pub async fn collect(
+    db_pool: deadpool_postgres::Pool,
+    auth_header: String,
+) -> Result<impl Reply, Rejection> {
+    let auth::Auth { username, password: _ } = auth::validate_header(auth_header.as_str())?;
+    let pool = db_pool
+        .get()
+        .await
+        .to_internal_error()?;
+
+    let user = crate::db::get_user_by_username(&pool, &username).await.to_internal_error()?;
+
+    let new_balance = user.balance + user.get_production() * user.collected_timestamp.duration_since(SystemTime::now()).to_internal_error()?.as_secs_f64();
+
+    let updated_user = crate::db::update_collect_user(&pool, &username, new_balance).await.to_internal_error()?;
+
+    Ok(warp::reply::json(&updated_user))
+}
+
+/// Upgrade the user's levels
+/// 
+/// This will return a 403 if the user does not have enough money to upgrade
+pub async fn upgrade(
+    db_pool: deadpool_postgres::Pool,
+    auth_header: String,
+) -> Result<impl Reply, Rejection> {
+    let auth::Auth { username, password: _ } = auth::validate_header(auth_header.as_str())?;
```

Comment:
i feel like this should be handled via the auth handler func, I'd rather it returned the user and then functions used that instead of every func calling validate_header manually

Reply:
True